omisego-sdk/src/main/java/co/omisego/omisego/websocket/OMGSocketClient.kt

```diff
@@ -0,0 +1,264 @@
+import okhttp3.WebSocket
+import okhttp3.logging.HttpLoggingInterceptor
+
+@Suppress("OVERRIDE_BY_INLINE")
+/**
+ * The [OMGSocketClient] represents an object that knows how to interact with the eWallet Web Socket API.
+ * An instance should be created by using [OMGSocketClient.Builder]. For example,
+ *
+ * <code>
+ *     val config = ClientConfiguration(
+ *          baseURL = "YOUR_BASE_URL",
+ *          apiKey = "YOUR_API_KEY",
+ *          authenticationToken = "YOUR_AUTH_TOKEN"
+ *      )
+ *
+ *     val omgSocketClient = OMGSocketClient.Builder{
+ *          clientConfiguration = config
+ *          debug = false
+ *     }.build()
+ * </code>
+ *
+ * The available methods and details are listed in the [SocketClientContract.Client]
+ *
+ * @see SocketClientContract.Client
+ */
+class OMGSocketClient internal constructor(
+    internal val okHttpClient: OkHttpClient,
+    internal var request: Request,
+    internal val socketSendParser: SocketClientContract.PayloadSendParser
+) : SocketClientContract.Client, SocketChannelContract.SocketClient {
+    internal var wsClient: WebSocket? = null
+    override lateinit var socketChannel: SocketClientContract.Channel
+
+    /**
+     * A [SocketHeartbeat] is responsible for schedule sending the heartbeat event for keep the connection alive
+     */
+    override val socketHeartbeat: SocketClientContract.SocketInterval by lazy {
+        SocketHeartbeat(SocketMessageRef().apply { scheme = SocketMessageRef.SCHEME_HEARTBEAT })
+    }
+
+    /**
+     * Immediately and violently release resources held by this web socket, discarding any enqueued
+     * messages. This does nothing if the web socket has already been closed or canceled.
+     */
+    override fun cancel() {
+        wsClient?.cancel()
+    }
+
+    /**
+     * A boolean indicating if all messages have been sent to the server.
+     *
+     * @return true, if all messages have been sent, otherwise false.
+     */
+    override fun hasSentAllMessages(): Boolean =
+        (wsClient?.queueSize() ?: 0L) == 0L
+
+    /**
+     * Joining a channel by the given [SocketTopic].
+     *
+     * @param topic The topic (channel) to which the event to be sent.
+     * @param payload (Optional) the additional data you might want to send bundled with the request.
+     * @param listener The event you want to receive for the specified [Channel].
+     * Be careful, the listener should be related to the topic, otherwise you won't receive any message.
+     * For example, if you are sending the topic begins with "transaction_request", then the listener must be the [SocketCustomEventCallback.TransactionRequestCallback] event.
+     *
+     * @see SocketCustomEventCallback
+     */
+    override fun <T : SocketCustomEventCallback> joinChannel(
+        topic: SocketTopic<T>,
+        payload: Map<String, Any>,
+        listener: T
+    ) {
+        with(socketChannel) {
+            join(topic.name, payload)
+            setCustomEventListener(listener)
```

Comment:
Shouldn't be that done before joining, to avoid a possible message loss between the two operations?

Reply:
Yep, make sense 👍

Reply:
I've refactored all callbacks to listeners 👉https://github.com/omisego/android-sdk/pull/24/commits/02cb88490c78481e74fd470c7eb199136d8b6292

Will look at the `delegation` and re-use the contracts. Thanks for recommend this 🙂